utils_linux.go

```diff
@@ -237,12 +236,16 @@ func isRootless(context *cli.Context) (bool, error) {
 		}
 		// nil b stands for "auto detect"
 	}
-	// Even if os.Geteuid() == 0, it might still require rootless mode,
-	// especially when running within userns.
-	// So we use system.GetParentNSeuid() here.
-	//
-	// TODO(AkihiroSuda): how to support nested userns?
-	return system.GetParentNSeuid() != 0 || system.RunningInUserNS(), nil
+	// We don't check EUID/UID and namespaces here. See https://github.com/opencontainers/runc/issues/1837
+	// * When runc is executed in userns via Docker-in-LXD, isRootless() returns false, because LXD would set $USER to "root".
+	// * When runc is executed in userns via rootless img, buildkit, and a few other tools, isRootless() returns true, because they don't change environment variables after unsharing the userns and mapping UID=0 to the current user.
+	// * When runc is executed as a regular user in the initial namespace, isRootless() returns true
+	// * When runc is executed as the root in the initial namespace, isRootless() returns false
+	// Corner cases:
+	// * When runc is executed in userns via Docker-in-rootless-Docker ("rootless dind"), as the root in the contaienr, isRootless() returns false, and unlikely to work. The dockerd in the container would need to specify runc --rootless explicitly in this case. (And user would need to launch dockerd with --rootless explicitly, probably)
+	// * When runc is executed in "rootless dind", as a non-root in the contaienr, isRootless() returns true and likely to work.
+	u := os.Getenv("USER")
```

Comment:
The `USER` environment variable seems to be set by the login process (see [man login(1)](http://man7.org/linux/man-pages/man1/login.1.html). Garden runs as server (i.e. it is not logging in) and therefore `USER` is not set when runc is invoked which makes `isRootless` return false regardless of whether we want rootless or not (`os.Getenv` returns an empty string if the var is not set)

In order to support both cases via checking the `USER` value only if it is set and fallback to the effective user ID if not, i.e. something like
```golang

func isRootless(context *cli.Context) (bool, error) {
  if context != nil {
      ...
  }

  user, userSet := os.LookupEnv("USER")
  if userSet {
    return user != "" && user != "root", nil
  }

  return system.GetParentNSeuid() != 0 || system.RunningInUserNS(), nil
}
```

What do you think?

cc @ostenbom @teddyking @julz

Reply:
@danail-branekov this looks like a safer version, I think it makes sense to fallback to the euid when `USER` is not set

Reply:
Thanks, but it doesn't fix Docker-in-LXD regression, because Docker-in-LXD executes runc without `$USER`, as euid=0 in userns, but it expects `isRootless()` to return false...

Reply:
> Garden runs as server

How do you run rootless garden as a server? via `systemd --user`?

`systemd --user` seems to set `$USER`, at least on Ubuntu 18.04?

```console
suda@suda-ws01:~$ systemctl show-environment
LANG=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

```console
suda@suda-ws01:~$ systemctl --user show-environment
HOME=/home/suda
LANG=en_US.UTF-8
LOGNAME=suda
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
SHELL=/bin/bash
USER=suda
XDG_RUNTIME_DIR=/run/user/1001
```

Reply:
We run garden with monit, which doesn't set `USER`. We also run the server as a user with maximum user id that does not exist in `/etc/passwd/` or elsewhere, so it would be best if we could run runc in rootless mode without setting `$USER`. On the other hand we can always run with the `--rootless` flag, so I guess this isn't too big of a problem for us. 

To clarify, in the issue you said `When runc is executed in userns via Docker-in-LXD, isRootless() returns false, because LXD would set $USER to "root".` - I don't understand in which case `$USER` *is* being set in Docker-in-LXD here?

Reply:
>because LXD would set $USER to "root”

sorry this was my wrong assumption

Reply:
>> because LXD would set $USER to "root”
> sorry this was my wrong assumption

So if Docker-in-LXD does NOT set $USER at all (i.e. $USER is either NOT on the environment, or set to en empty string when runc is run), then your PR would make `isRootless` return false, i.e. it would run rootful. Is that what you want to achieve? Is that what fixes the regression?

If yes, could you please update the comment in the code to something like
```
When runc is executed in userns via Docker-in-LXD, isRootless() returns false, because LXD would not set $USER.
``` 
so that we do not get confused?

Reply:
updated. also added EUID check

Reply:
LXD does not set anything at all unless explicitly instructed via environment variables. As soon as the basic container setup is done everything is in the hands of the binary that is executed. In the regular case for LXD and LXC this means a full-blown init system.

Reply:
> updated. also added EUID check

Oh, great! The EUID check makes our CF Garden acceptance tests quite happy 👍 

Thanks!